### PR TITLE
DQN

### PR DESCRIPTION
**Pull request by Het-Shah**
This is vanilla DQN. I just wanted the basic structure ready. I will add options for other variants as I proceed. Refer #36 

I was wondering if we should move `DQN_Mlp` and future implementations such as `DQN_CNN` to `actor_critic.py`. Right now it is the same dir as `dqn.py` inside `utils.py`

### CODE REVIEW

**Comment by Sharad24** [comment]
FYI your commits are not being registered under your name as you have not set your user.name and user.email in config

**Comment by Het-Shah** [comment]
LOL sorry, changed to windows recently.

File: jigglypuffRL/deeprl/dqn/utils.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        return self.layers(x)
+    
+    def act(self, state, epsilon):
+        if np.random.rand() > epsilon:
+            state   = Variable(torch.FloatTensor(state).unsqueeze(0), volatile=True)
+            q_value = self.forward(state)
+            action  = q_value.max(1)[1].data[0].item()
+        else:
+            action = random.randrange(self.env.action_space.n)
+        return action
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/dqn/utils.py:31]: @ajaysub110 We might be able to use MLPCritic for DQNs right? Because afaik its a Value function MLP which is what DQN requires. If that's the case, this might be redundant.

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        reward = Variable(torch.FloatTensor(reward))
+        done = Variable(torch.FloatTensor(done))
+
+        q_values = self.model(state)
+        next_q_values = self.model(next_state)
+        # print(f"q_values: {q_values}, next_q_values: {next_q_values}")
+        q_value = q_values.gather(1, action.unsqueeze(1)).squeeze(1)
+        next_q_value = next_q_values.max(1)[0]
+        expected_q_value = reward + self.gamma * next_q_value * (1 - done)
+        # print(f"q_value: {q_value}")
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/dqn/dqn.py:128]: Remove commented print statements

File: jigglypuffRL/deeprl/dqn/utils.py (hunk 1 above)
Comment (ajaysub110) [reply jigglypuffRL/deeprl/dqn/utils.py:31]: Yeah, had mentioned this in my review as well

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+import numpy as np
+
+import torch
+import torch.nn as nn
+import torch.optim as opt
+import torch.autograd as autograd
+from torch.autograd import Variable
+import torch.nn.functional as F
+
+from IPython.display import clear_output
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/dqn/dqn.py:13]: IPython?

File: jigglypuffRL/deeprl/dqn/utils.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    
+# TO DO add CNN Networks
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/dqn/utils.py:33]: This might affect #12 

We can probably avoid this for now and add it later since we don't really need CNNs until we support Image based envs e.g. Atari (#24 )

File: jigglypuffRL/deeprl/dqn/utils.py (hunk 1 above)
Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/utils.py:31]: yea initially thought of using MLPCritic but then thought to keep classes for policy gradient and DQN different. Will change it to MLPCritic

File: jigglypuffRL/deeprl/dqn/utils.py (hunk 4 above)
Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/utils.py:33]: Yea this is fine as of now

File: jigglypuffRL/deeprl/dqn/utils.py (hunk 1 above)
Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/utils.py:31]: I still want this to be different from `MLPCritic` class as this might cause confusion to the user. Can we not keep policy gradient and DQN implementations of the network different?

**Comment by Het-Shah** [comment]
Adding Double DQN also :P Please review

File: jigglypuffRL/deeprl/dqn/utils.py (hunk 1 above)
Comment (Sharad24) [reply jigglypuffRL/deeprl/dqn/utils.py:31]: Critic in the actor-critic setting is supposed to be a value function-based method. DQN formulation is pretty much a value function based algorithm. So, I'd still suggest using the MLPCritic methods otherwise it becomes a bit useless as an independent class.

Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/utils.py:31]: I will be having some modifications in this class itself when I will be having Duelling DQN, this is why i wanted it to be a different class.

Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/utils.py:31]: Moreover won't we be taking the action based on the q-value in DQN? I did not understand as to what should I pass as action in the MLPCritic method.

Comment (Zeus3101) [reply jigglypuffRL/deeprl/dqn/utils.py:31]: The Critic function will take a state-action pair while DQN will only take state to return different values of actions. Also, theoretically doesn't the Critic learn the advantage values while the Q function learn Q-values?They can't be made the same. Also, the value_hist should then become q_hist.

**Comment by Sharad24** [comment]
Some changes will probably be required soon based on a design @ajaysub110 is going to be suggesting. It looks like the current Base class definitions are a bit messy

**Comment by Het-Shah** [comment]
I don't think DQN depends on actor_critic though. But I will wait for the PR on the issue.

File: jigglypuffRL/deeprl/dqn/utils.py (hunk 1 above)
Comment (Sharad24) [reply jigglypuffRL/deeprl/dqn/utils.py:31]: You can use `MLPValue` from `Values.py`

**Comment by Sharad24** [comment]
Yup, try using `MLPValue` from `Values.py`? You'd need to add a method to get the epsilon greedy action there I guess.

**Comment by Sharad24** [comment]
On a side note, @ajaysub110 I think it would be a good idea to have support for both Actor-Critic definitions and Policy And Value Function classes.

**Comment by Het-Shah** [comment]
> I think it would be a good idea to have support for both Actor-Critic definitions and Policy And Value Function classes.

Won't this be redundancy?

**Comment by Sharad24** [comment]
Ideally, Actor definitions should be inheriting from Base Policies. In regards to Value functions, having support for them makes sense, since we have a clear case (DQN) where we need them separately.

Would it be a bad idea to inherit `BaseActor` from `BasePolicy` and `BaseCritic` from `BaseValue`?

**Comment by ajaysub110** [comment]
> On a side note, @ajaysub110 I think it would be a good idea to have support for both Actor-Critic definitions and Policy And Value Function classes. But use Actor Critic definitions wherever possible

I don't understand why that would be required. Can the Critic class not be used as Value also?

**Comment by Sharad24** [comment]
No, As pointed out, critic would take in an action as well as a state, whereas a value function will take in only a state observation

**Comment by Het-Shah** [comment]
Yea, I mean what I think is that we should be keeping policy gradient classes different from DQN classes

**Comment by Het-Shah** [comment]
https://github.com/Kaixhin/spinning-up-basic/blob/master/models.py
I had a look at what open ai has been using they have also kept actor-critic and DQN value function different

**Comment by ajaysub110** [comment]
But otherwise they're the same right? Then can't we just create one class and call it with different dimensions?

**Comment by Het-Shah** [comment]
Yea the idea is same for both, but why mix implementations. This will be bad for the readability wise.

**Comment by ajaysub110** [comment]
Okay cool. So we will have a Base Actor and Base Critic inheriting from BasePolicy and BaseValue right?

**Comment by Het-Shah** [comment]
I think we can still manage without BasePolicy and BaseValue. Both the BasePolicy and BaseActor seem to do the same thing logically to me. We are anyway having different implementation for DQN, we might not need BasePolicy and BaseValue anywhere.

**Comment by Sharad24** [comment]
Yeah, In my opinion, that seems like an optimal design as of now. 

If we go ahead with this, a thing here to take care of is to keep the Base classes as _inheritable_ as possible

**Comment by ajaysub110** [comment]
> Yeah, In my opinion, that seems like an optimal design as of now.
> 
> If we go ahead with this, a thing here to take care of is to keep the Base classes as _inheritable_ as possible

This is for only BaseValue and BasePolicy?

**Comment by Sharad24** [comment]
A reason why I'm not pushing for having a separate DQN class value function is, it breaks into the abstraction layer of Policies/Values/etc with a customized class. I'm not sure if this is best to have at the core of the library.

Value functions and Policy functions come to me as a more general concept in RL (applicable to almost every algo), which is why I'd be willing to keep them with support for ActorCritic. Sure they may not defer right now much implementation-wise. 

Spinning Up (or even Stable Baselines) is an implementation first sort of library. So their focus is not much on building abstractions that are helpful and reusable but rather just supporting high-level implementations. If they need another algorithm tomorrow, they'd just implement another class for that so that they can call a single function. There's not really any focus on building generally usable abstractions. Which is why I think we should support Policy and Value including an Actor-Critic framework since its used a lot

This is definitely a very big and important topic of debate, so let me know if you disagree anywhere.

**Comment by ajaysub110** [comment]
Completely agree. So let's just have (as of now) a BasePolicy, BaseValue which will be general to all deeprl algos. We will then have BaseActorCritic whose subclasses group inherited instances of BasePolicy and BaseValue. Does this sound okay?

**Comment by Het-Shah** [comment]
Yea cool, I agree now. Had a bit of confusion earlier as to where this was going.

**Comment by Het-Shah** [comment]
I will be using MLPValue (modifying to return an action based on e-greed) instead of having another implementation, that is fine right?

**Comment by ajaysub110** [comment]
I'll open a draft PR now for the actor_critic issue so that you can refer to the new policy and value classes there

**Comment by Het-Shah** [comment]
Closing this
